src/zone.c

```diff
@@ -115,89 +115,28 @@ diagnostic_push()
 msvc_diagnostic_ignored(4996)
 
 #if _WIN32
-nonnull_all
-static bool is_separator(int c)
-{
-  return c == '\\' || c == '/';
-}
-
-nonnull_all
-static bool is_rooted(const char *s)
-{
-  if ((s[0] >= 'A' && s[0] <= 'Z') || (s[0] >= 'a' && s[0] <= 'z'))
-    return s[1] == ':';
-  return false;
-}
-
-nonnull_all
-static bool is_relative(const char *s)
-{
-  // rooted paths can be relative, e.g. C:foo
-  if (is_rooted(s))
-    return !is_separator((unsigned char)s[2]);
-  if (is_separator((unsigned char)s[0]))
-    return !(s[1] == '?' || is_separator((unsigned char)s[1]));
-  return false;
-}
-
 // The Win32 API offers PathIsRelative, but it requires linking with shlwapi.
 // Rewriting a relative path is not too complex, unlike correct conversion of
 // Windows paths in general (https://googleprojectzero.blogspot.com/2016/02/).
 // Rooted paths, relative or not, unc and extended paths are never resolved
 // relative to the includer.
-nonnull((2,3))
-static int32_t resolve_path(
-  const char *includer, const char *include, char **path)
+nonnull_all
+static int32_t resolve_path(const char *include, char **path)
 {
-  // support relative non-rooted paths only
-  if (*includer && is_relative(include) && !is_rooted(include)) {
-    assert(!is_relative(includer));
-    char buffer[16];
-    int length = snprintf(
-      buffer, sizeof(buffer), "%s/%s", includer, include);
-    if (length < 0)
-      return ZONE_OUT_OF_MEMORY;
-    char *absolute;
-    if (!(absolute = malloc(length + 1)))
-      return ZONE_OUT_OF_MEMORY;
-    (void)snprintf(
-      absolute, (size_t)length + 1, "%s/%s", includer, include);
-    *path = _fullpath(NULL, absolute, 0);
-    free(absolute);
-  } else {
-    *path = _fullpath(NULL, include, 0);
-  }
-
-  if (*path)
+  if ((*path = _fullpath(NULL, include, 0)))
     return 0;
   return (errno == ENOMEM) ? ZONE_OUT_OF_MEMORY : ZONE_NOT_A_FILE;
 }
 #else
 nonnull_all
-static int32_t resolve_path(
-  const char *includer, const char *include, char **path)
+static int32_t resolve_path(const char *include, char **path)
 {
   char *resolved;
   char buffer[PATH_MAX + 1];
 
-  if (*includer && *include != '/') {
-    assert(*includer == '/');
-    int length = snprintf(
-      buffer, sizeof(buffer), "%s/%s", includer, include);
-    if (length < 0)
-      return ZONE_OUT_OF_MEMORY;
-    char *absolute;
-    if (!(absolute = malloc((size_t)length + 1)))
-      return ZONE_OUT_OF_MEMORY;
-    (void)snprintf(
-      absolute, (size_t)length + 1, "%s/%s", includer, include);
-    resolved = realpath(absolute, buffer);
-    free(absolute);
-  } else {
-    resolved = realpath(include, buffer);
-  }
+  resolved = realpath(include, buffer);
 
-  if (!resolved)
+  if (!(resolved = realpath(include, buffer)))
```

Comment:
Since you strlen, malloc and memcpy the buffer anyway, would it perhaps be better to call realpath(3) with `buffer = NULL`, thereby letting the standard library to do the heavy lifting?

> If resolved_path is specified as NULL, then realpath() uses malloc(3) to allocate a buffer of up to PATH_MAX bytes to hold the resolved pathname, and returns a pointer to this buffer.  The caller should deallocate this buffer using free(3).

Reply:
We had that at first, but not all `realpath` implementations (specifically Solaris 10) handles that well. I agree it'd be easier, but let's leave it as is for now.

Reply:
> not all realpath implementations (specifically Solaris 10) handles that well

Wow, is anybody still running that in production? :-O

In that case, I guess it makes sense. I believe the current code is correct, though I would probably use strdup/strndup myself. After all, it exists precisely to avoid possible (future) errors caused by hand-copying buffers.

Btw realpath(3) mentions that Solaris uses a different constant for the maximum path length (MAXPATHLEN). The correctness of the realpath call stands on that (since realpath assumes the buffer is at least PATH_MAX long) and I have no experience with Solaris at all, so I'm just letting you know.

Reply:
I do not know to which extent it's used in production, but maintainers from [OpenCSW](https://www.opencsw.org/) did notify us of NSD 4.10 (RC1?) not compiling. I fixed and tested it to help them out, but I agree it's not the youngest operating system out there :slightly_smiling_face:.